---
title: "Evaluation of Nanocoding Multiplexing for Six-plex Mouse Spleen Sample"
output: html_document
date: "2024-08-05"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Six-plex Mouse Spleen Samples Test

This vignette is to show the minimal data processing steps after CellRanger output (shown in folder 10X_CellRangerReport/), in order to repeat major results shown in the Figure 3 for six-plex mouse spleen test. This script includes 3 sections: \
1. Preprocessing of RNA matrix for cell quality evaluation;\
2. Barcode matrix analysis and cell classifications by HTO barcodes using CellHashR; \
3. Automatic cell type classification by RNA profile and differential gene expression between barcode identified groups. \

#Preprocessing RNA matrix
First, load necessary packages:

```{r,message=FALSE,warning=FALSE}

library(tidyverse)
library(Seurat)
library(dittoSeq)
library(sctransform)
library(ggplot2)
library(cellhashR)
library(celldex)
library(scran)
library(scRNAseq)
library(SingleR) 
library(scater)

```

Then, load RNA matrix, evaluate cell quality: 

```{r pressure, echo=TRUE, fig.width=10, fig.height=5}
#load matrix. This data is available in our GEO deposit (acession number:).
scData<-Read10X(data.dir="~/Spleen_sixplex/cellranger/filtered_feature_bc_matrix")

#the output after CellRanger count
#subset RNA matrix
sp6<-CreateSeuratObject(
  counts=scData[[1]],
  names.field=1,
  names.delim = "-",
  min.cells = 20)

#Add one column in metadata of mitocondrial gene percentage in each cell
mgenes <-  rownames(sp6) 
MT.genes.mouse <- grep("^mt-", mgenes, value = TRUE)
sp6[["percent.mt"]] <- PercentageFeatureSet(sp6, features = MT.genes.mouse) 


#Barcode matrix is subset to another Seurat Obeject
sp6_barcode<-CreateSeuratObject(
  counts=scData[[2]],
  names.field=1,
  names.delim = "-",
  min.cells = 20)
```

RNA quality evaluation: 
```{r pressure, echo=TRUE, fig.width=20, fig.height=5}
plot1 <- FeatureScatter(sp6, 
                        feature1 = "nCount_RNA", 
                        feature2 = "percent.mt") 
plot2<-  FeatureScatter(sp6, 
                        feature1 = "nCount_RNA", 
                        feature2 = "nFeature_RNA")
plot3 <-  FeatureScatter(sp6, 
                        feature1 = "percent.mt", 
                        feature2 = "nFeature_RNA")
plot1 + plot2 + plot3

```
Then, we filter cells with high percentage of mitochondrial gene expression. The cells to be removed are defined by the percent.mt value that have 3 times higher expression than median absolute deviation (MAD), using scuttle_1.12.0 package. 
```{r pressure, echo=TRUE}

temp1 <- scater::isOutlier(sp6$percent.mt, 
                           nmads = 3, type = "higher")
table(temp1)
mt.threshold <- min(sp6$percent.mt[temp1]) 
sp6_filter <- subset(sp6, 
                        subset = percent.mt < mt.threshold)
ncol(sp6_filter)
```
Then we normalize RNA assay using methods from Seurat package. 
```{r pressure, echo=TRUE, message=FALSE, warning=FALSE}

sp6_filter <- SCTransform(sp6_filter, 
                             method = "glmGamPoi",
                             vars.to.regress = "percent.mt", 
                             return.only.var.genes = FALSE, 
                             vst.flavor = "v2")
sp6_filter <- RunPCA(sp6_filter, verbose = FALSE) 
ElbowPlot(sp6_filter, 50)  

# UMAP plot generated using specified number of dimensions based on elbow plot:
sp6_filter <- RunUMAP(sp6_filter, 
                         dims = 1:40, 
                         verbose = FALSE) 
```
Cell clustering and UMAP plot: 
```{r pressure, echo=TRUE, fig.width=10, fig.height=10}
sp6_filter <- FindNeighbors(sp6_filter, dims = 1:40, verbose = FALSE) 
sp6_filter <- FindClusters(sp6_filter, verbose = FALSE)

table(sp6_filter$seurat_clusters) 
dittoDimPlot(sp6_filter, 
             var = "seurat_clusters", 
             reduction.use = "umap", 
             do.label = TRUE, 
             labels.highlight = FALSE)
```

#Barcode Matrix Analysis 
 In this section, we work on the barcode matrix and evaluate its quality using cellhashR package. 
 The original bimodal distribution of barcode count hitograms can be found in 10X_CellRangerReports/ folder. 
```{r pressure, echo=TRUE, fig.width=10, fig.height=10}

#subset the barcode matrix to match filtered cell from RNA matrix:
joint.bcs<-intersect(colnames(sp6_filter),colnames(sp6_barcode)) 
sp6_barcode<- sp6_barcode[,joint.bcs] 

raw_counts <- GetAssayData(sp6_barcode, assay = "RNA", layer  = "counts") 
raw_counts_matrix <- as.matrix(raw_counts) 
write.csv(raw_counts_matrix, file = "raw_counts.csv")

barcodeData <- ProcessCountMatrix(rawCountData = 'raw_counts.csv', minCountPerCell = 0) 

#the barcode distribution and classification results are generated by using multiseq or htodemux method 
calls <- GenerateCellHashingCalls(barcodeMatrix = barcodeData, methods = c('multiseq', 'htodemux')) 

#add the classification results to the metadata of sp6_filter seurat object 
sp6_filter[["deMULTIplex"]]<-calls$multiseq
sp6_filter[["HTOdemux"]]<-calls$htodemux



``` 

#Cell Clustering and Barcode Identified Group Based Analysis 

In this section, we annotated the cell type based on ImmGen database using SingleR package. Then, by adding the cell-sample information obtained from barcode matrix, we get the percentage of each cell type in each sample. At the end, we compared the differential gene expression features in the same cell type between lean or obese mice. The difference is evaluated through both single-cell level gene expression, and bulk-level gene expression using pseudobulk analysis, using Seurat package. 

```{r pressure, echo=TRUE, fig.width=10, fig.height=5}
ImmGen <- celldex::ImmGenData()  

Idents(sp6_filter)<-"seurat_clusters"

singleR <- as.SingleCellExperiment(sp6_filter, assay= "SCT") 

singler.ImmGen <- SingleR(test = singleR, ref = ImmGen, labels = (ImmGen$label.main))

#add the ImmGen based cell type annotation to metadata of sp6_filter
sp6_filter[["ImmGen"]] <- singler.ImmGen$labels 

SR1 <- dittoDimPlot(sp6_filter, var=sp6_filter$seurat_clusters,reduction.use = "umap",do.label=TRUE) 
SR2 <- dittoDimPlot(sp6_filter, var=sp6_filter$ImmGen,reduction.use = "umap",do.label=TRUE, labels.highlight = FALSE) 

SR1+SR2

```

```{r pressure, echo=TRUE, fig.width=10, fig.height=5}
#subset Monocytes for differential Gene analysis 

Idents(sp6_filter)<-"ImmGen"
sp6_filter_Monocytes<-subset(sp6_filter, idents="Monocytes")

Idents(sp6_filter_Monocytes)<-"HTOdemux"
DE_Mono <- FindMarkers(sp6_filter_Monocytes, 
                        ident.1 = c("HTO1","HTO2","HTO3"), 
                        ident.2 = c("HTO4","HTO5","HTO6"),
                        assay = "SCT", 
                        slot = "data",
                        min.pct = 0.25, 
                        logfc.threshold = 0.25)

DE_Mono$gene <- rownames(DE_Mono) 
DE_Mono<- DE_Mono %>% mutate(logP = -log10(p_val), signif = ifelse(p_val < 0.01&abs(avg_log2FC)>1, "significant", "not significant"))

Mon_genes_to_label <- DE_Mono %>% 
  filter(signif== "significant")

p<-sggplot(DE_Mono, aes(x = avg_log2FC, y = logP)) +
  geom_point(aes(color = signif), alpha = 0.5) +
  geom_text(data = Mon_genes_to_label, aes(label = gene), vjust = 1.5, hjust = 0.5, check_overlap = TRUE, size = 3) +
  theme_minimal() +
  labs(title = "Volcano plot of differentially expressed genes in Monocytes", x = "Log2 Fold Change", y = "-Log10 p-value") +
  scale_color_manual(values = c("significant" = "red", "not significant" = "black"))
p

#then we run pseudobulk analysis to include biological differences in each individual 
Idents(sp6_filter)<-"HTOdemux" 
sp6_filter$weight<-"Default"

sp6_filter$weight[sp6_filter$HTOdemux%in%c("HTO1","HTO2","HTO3")]<-"Obese" 
sp6_filter$weight[sp6_filter$HTOdemux%in%c("HTO4","HTO5","HTO6")]<-"Lean"

sp6_filter_singlets <- subset(sp6_filter, subset = weight != "Default") 
pseudo_sp6 <- AggregateExpression(sp6_filter_singlets, assays = "SCT", return.seurat = T, group.by = c("weight", "HTOdemux", "ImmGen"))

pseudo_sp6$celltype.stim <- paste(pseudo_sp6$weight, pseudo_sp6$ImmGen, sep = "_") 
Idents(pseudo_sp6) <- "celltype.stim"
bulk.mono.de <- FindMarkers(object = pseudo_sp6, 
                            ident.1 = "Obese_Monocytes", 
                            ident.2 = "Lean_Monocytes",
                            test.use = "DESeq2")
head(bulk.mono.de, n = 15)

sp6_filter_singlets$weight_celltype <- paste0(sp6_filter_singlets$weight, "_", sp6_filter_singlets$ImmGen)
Idents(sp6_filter_singlets) <- "weight_celltype"
VlnPlot(sp6_filter_singlets, features = "Plin2", idents = c("Obese_Monocytes", "Lean_Monocytes"), group.by = "HTOdemux", ncol = 1)  

```

##Seesion Info
```{r pressure, echo=TRUE}
sessionInfo()
```
